app/user.go

```diff
@@ -2141,3 +2141,39 @@ func (a *App) invalidateUserCacheAndPublish(userId string) {
 func (a *App) GetKnownUsers(userID string) ([]string, *model.AppError) {
 	return a.Srv().Store.User().GetKnownUsers(userID)
 }
+
+// ConvertBotToUser converts a bot to user.
+func (a *App) ConvertBotToUser(bot *model.Bot, userPatch *model.UserPatch, sysadmin bool) (*model.User, *model.AppError) {
+	user, err := a.Srv().Store.User().Get(bot.UserId)
+	if err != nil {
+		return nil, err
+	}
+
+	if sysadmin && !user.IsInRole(model.SYSTEM_ADMIN_ROLE_ID) {
+		if _, err = a.UpdateUserRoles(
+			user.Id,
+			fmt.Sprintf("%s %s", user.Roles, model.SYSTEM_ADMIN_ROLE_ID),
+			false); err != nil {
+			return nil, err
+		}
+	}
+
+	user.Patch(userPatch)
+
+	user, err = a.UpdateUser(user, false)
+	if err != nil {
+		return nil, err
+	}
+
+	err = a.UpdatePassword(user, *userPatch.Password)
+	if err != nil {
+		return nil, err
+	}
+
+	err2 := a.Srv().Store.Bot().PermanentDelete(bot.UserId)
+	if err2 != nil {
```

Comment:
To avoid having a functional bot or some inconsistency, maybe we can retry using a soft deletion?

Reply:
Do you mean disabling the bot?

Reply:
Yes, at least to inactivate it to avoid weird behavior (just an idea)

Reply:
I think it's fine to stay as is because it will return an error nonetheless.